### ISSUE TITLE
FirebaseUI not showing up half of the time

### ISSUE THREAD

**Issue by gregleleu** [JohnCoene/firebase#30]
I have to reload the page multiple times for FirebaseUI to appear, when I reload more it goes away again, then back...

When it doesn't appear: FirebaseUI$launch is called before the `Shiny.addCustomMessageHandler`for "fireblaze-ui-config" is registered (I see it by logging both events to the js console).

Any idea how to fix the timing?

**Comment by gregleleu** [comment]
For the moment I have created an input showing the ui was created
```
Shiny.setInputValue("uistarted", true); // just after the addCustomMessageHandler is ran
```
and wrapped my launch command in 
```
observeEvent(input$uistarted, {if (input$uistarted) f$launch(flow = "popup")}, ignoreNULL = TRUE, ignoreInit = TRUE)
````
It works, but it's not pretty

**Comment by gregleleu** [comment]
Spending a bit more time on this, I'm thinking this solution actually has some merit: everything else in the server function that talks to the javascript is wrapped in observers/reactives.
Just wrapping `launch` in an observer seems to work as well, and it doesn't require an extra input value etc.
```
observe({f$launch(flow = "popup")})
```
And from the reactlog, it's never called again.

What do you think? Am I the only one with this issue?

**Comment by JohnCoene** [comment]
I'm wondering if it has to do with the event cycle/reactive. Which version of shiny are you using?

**Comment by gregleleu** [comment]
Latest CRAN: 1.7.1
Re. my previous comment: I couldn't find other example of code which interacts with the javascript side and which is not wrapped in a reactive context

**Comment by JohnCoene** [comment]
It's fair to say that it's generally the case. I'm not sure how to integrate this nicely in the app, do you have any suggestion?

**Comment by gregleleu** [comment]
Did a bit more reading: the server and the client are never going to be synchronous, and anyway we don't want to "block" the execution of the server function until firebase is ready to launch. 

Therefore the idea is to run `f$launch()` in a reactive context. Two ways from what I can tell:
1.  `observe({f$launch(flow = "popup")})`: creates an observer, it is run once when Shiny initializes all observers after the UI is loaded (from what I understand), the observer stays but is never called again
2. `session$onFlush(function() {f$launch(flow = "popup")}, once = TRUE)`: tells shiny to run that function once before the first initialization of the inputs

I like option 1 better; it is more common Shiny, but it leaves a useless observer around (I don't think it really matters). Option 2 is a bit verbose.

Anyway, I think it's just a documentation change – nothing to change in the code – and maybe a bit of testing (only tested my app). I'm only familiar with the FirebaseUI flow, I have no idea how this maps to the other options.

**Comment by JohnCoene** [comment]
Agreed, I'll change the documentation when I find the time, thank you again for looking into this!
